pkg/kubectl/cmd/top.go

```diff
@@ -0,0 +1,166 @@
+	"k8s.io/kubernetes/pkg/kubectl"
+	cmdutil "k8s.io/kubernetes/pkg/kubectl/cmd/util"
+	client "k8s.io/kubernetes/pkg/client/unversioned"
+	"k8s.io/kubernetes/pkg/api"
+	"k8s.io/kubernetes/pkg/api/unversioned"
+	"k8s.io/kubernetes/pkg/kubectl/metricsutil"
+)
+
+// TopOptions is the start of the data required to perform the operation. As new fields are added, add them here instead of
+// referencing the cmd.Flags()
+type TopOptions struct {
+}
+
+var (
+	topLong = dedent.Dedent(`
+		Display Resource (CPU/Memory/Storage) usage of nodes or pods.
+
+		The top command allows you to see the resource consumption of the nodes or pods.
+		It downloads the usage metrics of a given resource (node/pod) via the Resource Metrics API.
+		`)
+
+	topExample = dedent.Dedent(`
+		  # Show metrics for all nodes in the default namespace
+		  kubectl top node
+
+		  # Show metrics for all nodes in the given namespace
+		  kubectl top node --namespace=NAMESPACE
+
+		  # Show metrics for a given pod in the default namespace
+		  kubectl top pod POD_NAME
+
+		  # Show metrics for a given pod and its containers
+		  kubectl top pod POD_NAME --containers
+
+		  # Show metrics for the pods defined by the selector query
+		  kubectl top pod --selector="KEY: VALUE"`)
+)
+
+var HandledResources []unversioned.GroupKind = []unversioned.GroupKind{
+	api.Kind("Pod"),
+	api.Kind("Node"),
+};
+
+var GetResourcesHandledByTop = func() []string {
+	keys := make([]string, 0)
+	for _, k := range HandledResources {
+		resource := strings.ToLower(k.Kind)
+		keys = append(keys, resource)
+	}
+	return keys
+}
+
+func NewCmdTop(f *cmdutil.Factory, out io.Writer) *cobra.Command {
+	options := &TopOptions{}
+	validArgs := GetResourcesHandledByTop()
+	argAliases := kubectl.ResourceAliases(validArgs)
+
+	cmd := &cobra.Command{
+		Use:     "top TYPE [NAME] [flags]",
+		Short:   "Display Resource (CPU/Memory/Storage) usage of nodes or pods",
+		Long:    topLong,
+		Example: topExample,
+		Run: func(cmd *cobra.Command, args []string) {
+			err := RunTop(f, out, cmd, args, options)
+			cmdutil.CheckErr(err)
+		},
+		ValidArgs:  validArgs,
+		ArgAliases: argAliases,
+	}
+	cmd.Flags().StringP("selector", "l", "", "Selector (label query) to filter on")
+	cmd.Flags().Bool("containers", false, "If present, print usage of containers within a pod. Ignored if the requested resource type is \"node\".")
+	// TODO: add support for --all-namespaces
+	cmdutil.AddInclude3rdPartyFlags(cmd)
+	return cmd
+}
+
+func RunTop(f *cmdutil.Factory, out io.Writer, cmd *cobra.Command, args []string, options *TopOptions) error {
+	if len(args) == 0 {
+		fmt.Fprint(out, "You must specify the type of resource to get.")
+		return cmdutil.UsageError(cmd, "Required resource not specified.")
+	}
+	selector := cmdutil.GetFlagString(cmd, "selector")
+	params := map[string]string{"labelSelector": selector}
+	printContainers := cmdutil.GetFlagBool(cmd, "containers")
+	cmdNamespace, _, err := f.DefaultNamespace()
```

Comment:
I'm not familiar with kubect code, but what does DefaultNamespace return?
Isn't it default namespace? If so, we should probably return the namespace that was passed by the flag...